README.md
# FNFTpy - a wrapper for FNFT

This module provides a python interface (wrapper functions) for [FNFT](https://github.com/FastNFT), 
a library for the numerical computation of nonlinear Fourier transforms.

For FNFTpy to work, a copy of FNFT has to be installed.
For general information, source files and installation of FNFT,
visit FNFT's github page: [https://github.com/FastNFT](https://github.com/FastNFT)


## current state - access functions from FNFT version 0.5.0

for changes and latest updates see [Changelog](CHANGELOG.md)


### Korteweg-de-Vries equation with vanishing boundary conditions:

* calculates the Nonlinear Fourier Transform of the Korteweg-de-Vries equation
<img src="https://render.githubusercontent.com/render/math?math=q_x%2B 6qq_t %2B q_{ttt} =0">
<!--$$q_x + 6qq_{t} + q_{ttt}=0$$-->
with vanishing boundary conditions

* calculation of the continuous spectrum:
	- reflection coefficient and/or the NFT coefficients a and b
* calculation of the discrete spectrum:
	- bound states 
	- norming constants and/or residues 
	- initial guesses for bound states may be provided

 
Function **kdvv**: 

* easy-to-use Python function, options can be passed as optional arguments 
* minimal example:

```python
import numpy as np
from FNFTpy import kdvv
D = 256
tvec = np.linspace(-1, 1, D)
q = np.zeros(D, dtype=np.complex128)
q[:] = 2.0 + 0.0j
gs = np.sqrt(np.max(np.abs(q))) / 1000.  # grid spacing
Xi1 = -2
Xi2 = 2
M = 8  # points for continuous spectrum
K = 8  # number of expected bound states
Xivec = np.linspace(Xi1, Xi2, M)
res = kdvv(q, tvec, K, M, Xi1=Xi1, Xi2=Xi2, gs=gs)
print("\n----- options used ----")
print(res['options'])
print("\n------ results --------")
print("FNFT return value: %d (should be 0)" % res['return_value'])
print("continuous spectrum: ")
for i in range(len(res['cont_ref'])):
    print("%d : Xi=%.3f   %.3e + %.3ej"%(i,
            Xivec[i], np.real(res['cont_ref'][i]),
            np.imag(res['cont_ref'][i])))
print("discrete spectrum")
for i in range(res['bound_states_num']):
    print("%d : bound state  %.3e + %.3ej with norming const %.3e + %.3ej"% (i,
              np.real(res['bound_states'][i]),np.imag(res['bound_states'][i]),
              np.real(res['disc_norm'][i]),np.imag(res['disc_norm'][i])))
```

* for full description call `help(kdvv)`
  
* function **kdvv_wrapper**:
	* mimics the function fnft_kdvv from FNFT.
	* for full description call ```help(kdvv_wrapper)```
  
### Manakov equation with vanishing boundary conditions


* calculates the Nonlinear Fourier Transform of the Manakov equation 
<img src="https://render.githubusercontent.com/render/math?math=iq_x %2B q_{tt} \pm 2q|q|^2=0, \quad  q=[q1,q2]">
<!--$$ iq_x + q_{tt} \pm 2q|q|^2=0, \quad  q=[q1; q2]$$-->
with vanishing boundary conditions

* calculation of the continuous spectrum:
	- reflection coefficient and/or the NFT coefficients a, b1 and b2
* calculation of the discrete spectrum:
	- bound states 
	- norming constants and/or residues 
	- initial guesses for bound states may be provided
	

* Function **manakovv**: 
	* easy-to-use Python function, options can be passed as optional arguments 
  * minimal example:
```python
import numpy as np
from FNFTpy import manakovv
D= 1024
M = 8
Xi1 = -1.75
Xi2 = 2
kappa = 1
q1 = np.zeros(D, dtype=np.complex128)
q2 = np.zeros(D, dtype=np.complex128)
tvec = np.linspace(-2, 2, D)
# standard
q1[:] = 2.0 + 0.0j
q2[:] = .650 + 0.0j
Xivec = np.linspace(Xi1, Xi2, M)
res = manakovv(q1, q2, tvec, M=M,  Xi1=Xi1, Xi2=Xi2)
print("-- continuous spectrum ---\n")
print("                  \t part 1\t\t\t\t\t\t part2")
for i in range(M):
    print("%d Xi = %.5f  \t%.4e + %.4ei \t%.4e + %.4ei" % (i + 1, Xivec[i],
                                       np.real(res['cont_ref1'][i]), np.imag(res['cont_ref1'][i]),
                                       np.real(res['cont_ref2'][i]), np.imag(res['cont_ref2'][i])))
```
	* for full description call `help(manakovv)`
        
* Function **manakovv_wrapper**:
    * mimics the function fnft_manakovv from FNFT.
    * for full description call ```help(manakovv_wrapper)```
  
### Nonlinear Schroedinger Equation with periodic boundary conditions

* calculation of the  Nonlinear Fourier Transform of the Nonlinear Schroedinger equation
<img src="https://render.githubusercontent.com/render/math?math=iq_x %2B q_{tt} \pm 2q|q|^2=0">
<!--$$ iq_x + q_{tt} \pm 2q|q|^2=0$$--> 
with (quasi-)periodic boundary conditions


  * The main and auxiliary spectra can be calculated.
  * Function **nsep**: 
    * easy-to-use Python function, options can be passed as optional arguments 
  * minimal example:

```python
import numpy as np
from FNFTpy import nsep
print("\n\nnsep example")
# set values
D = 256
tvec = np.linspace(0, 2*np.pi, D)
q = np.exp(2.0j * tvec)
# call function
res = nsep(q, 0, 2 * np.pi, bb=[-2, 2, -2, 2], filt=1, kappa=1)
# print results
print("\n----- options used ----")
print(res['options'])
print("\n------ results --------")
print("FNFT return value: %d (should be 0)" % res['return_value'])
print("number of samples: %d" % D)
print('main spectrum')
for i in range(res['K']):
  print("%d :  %.6f  %.6fj" % (i, np.real(res['main'][i]),
                                   np.imag(res['main'][i])))
print('auxiliary spectrum')
for i in range(res['M']):
  print("%d :  %.6f  %.6fj" % (i, np.real(res['aux'][i]),
                                    np.imag(res['aux'][i])))
```
  * for full description call `help(nsep)`
  * Function **nsep_wrapper**:
    * mimics the function fnft_nsep from FNFT.
    * for full description call ```help(nsep_wrapper)```
     
  
### Nonlinear Schroedinger Equation with vanishing boundary conditions:


* calculates the Nonlinear Fourier Transform of the Nonlinear Schroedinger equation 
<img src="https://render.githubusercontent.com/render/math?math=iq_x %2B q_{tt} \pm 2q|q|^2=0">
<!--$$iq_x + q_{tt} \pm 2q|q|^2=0$$-->
with vanishing boundary conditions

* calculation of the continuous spectrum:
	- reflection coefficient and/or the NFT coefficients a and b
* calculation of the discrete spectrum:
	- bound states 
	- norming constants and/or residues 
	- initial guesses for bound states may be provided


* Function **nsev**:
	* easy-to-use Python function, options can be passed as optional arguments     
    * minimal example:
  
```python
import numpy as np
from FNFTpy import nsev

# set values
D = 256
tvec = np.linspace(-1, 1, D)
q = np.zeros(len(tvec), dtype=np.complex128)
q[:] = 2.0 + 0.0j
M = 8
Xi1 = -2
Xi2 = 2
Xivec = np.linspace(Xi1, Xi2, M)

# call function
res = nsev(q, tvec, M=M, Xi1=Xi1, Xi2=Xi2)

# print results
print("\n----- options used ----")
print(res['options'])
print("\n------ results --------")

print("FNFT return value: %d (should be 0)" % res['return_value'])
print("continuous spectrum")
for i in range(len(res['cont_ref'])):
    print("%d :  Xi = %.4f   %.6f  %.6fj" % (i, Xivec[i], np.real(res['cont_ref'][i]), np.imag(res['cont_ref'][i])))
print("discrete spectrum")
for i in range(len(res['bound_states'])):
    print("%d : %.6f  %.6fj with norming const %.6f  %.6fj" % (i, np.real(res['bound_states'][i]),
                                                             np.imag(res['bound_states'][i]),
                                                             np.real(res['disc_norm'][i]),
                                                             np.imag(res['disc_norm'][i])))
```        

  * for full description call `help(nsev)`
        
* Function **nsev_wrapper**:
	* mimics the function fnft_nsev from FNFT.
    * for full description call ```help(nsev_wrapper)```
     
### Nonlinear Schroedinger Equation with vanishing boundary conditions (Inverse Transformation): 
 * Perform the Inverse Nonlinear Fourier transform: the temporal field is calculated from the nonlinear spectrum.
 * the continuous part and (optional) the discrete part of the spectrum can be given.  

 * function **nsev_inverse**
    * easy-to-use Python function, options can be passed as optional arguments 
    * minimal example:

```python
from FNFTpy import nsev_inverse, nsev_inverse_xi_wrapper
import numpy as np
D = 1024
M = 2 * D
Tmax = 15
tvec = np.linspace(-Tmax, Tmax, D)
# calculate suitable frequency bonds (xi)
rv, xi = nsev_inverse_xi_wrapper(D, tvec[0], tvec[-1], M)
xivec = xi[0] + np.arange(M) * (xi[1] - xi[0]) / (M - 1)
# analytic field: chirp-free N=2.2 Satsuma-Yajima pulse
q = 2.2 / np.cosh(tvec)
# semi-analytic nonlinear spectrum
bound_states = np.array([0.7j, 1.7j])
disc_norming_const_ana = [1.0, -1.0]
cont_b_ana = 0.587783 / np.cosh(xivec * np.pi) * np.exp(1.0j * np.pi)
# call the function
res = nsev_inverse(xivec, tvec, cont_b_ana, bound_states, disc_norming_const_ana, cst=1, dst=0)
# compare result to analytic function
print("\n\nnsev-inverse example: Satsuma-Yajima N=2.2")
print("Difference analytic - numeric: sum((q_ana-q_num)**2) = %.2e  (should be approx 0) " % np.sum(
    np.abs(q - res['q']) ** 2))
``` 

  * Function **nsev_inverse_wrapper**:
    * mimics the function fnft_nsev_inverse from FNFT.
    * for full description call ```help(nsev_inverse_wrapper)```
# Requirements
 * Python 3.8 and above
 * additional Python module: NumPy (python-numpy)  
 
# Setup
 * you may install FNFTpy locally using **pip**:
   From within the project root folder run
     ```
     pip install .     # Install system wide
     pip install -e .  # Install in editable/development mode
     ```

 * **alternatively**, you may add the FNFTpy folder to your Python path.

 * Naturally, you need a compiled version of the FNFT C-library. See the
  [documentation for FNFT](https://github.com/FastNFT/FNFT) on how to build 
   the library on your device. 
 * FNFTpy needs to know where the C-library is located. 
   This configuration can be done by editing the function get_lib_path()
   in auxiliary.py. 
   
   Example:
 
```python
def get_lib_path():
    """Return the path of the FNFT file.

    Here you can set the location of the compiled library for FNFT.
    See example strings below.

    Returns:

    * libstring : string holding library path

    Example paths:

        * libstr = "C:/Libraries/local/libfnft.dll"  # example for windows            
        * libstr = "/usr/local/lib/libfnft.so"  # example for linux

    """
    libstr = "/usr/local/lib/libfnft.so"  # example for linux
    return libstr
```
    
 # License 
  
FNFTpy is provided under the terms of the [GNU General Public License, version 2.](https://www.gnu.org/licenses/old-licenses/gpl-2.0.html)

# Contact and contributors

* for bug reports, please use the github issue tracker

* contributors: 

    * Christoph Mahnke (chmhnk_at_googlemail_dot_com)
    * Shrinivas Chimmalgi
    * Simone Gaiarin / simgunz [https://github.com/simgunz]

CHANGELOG.md
## changelog

## 0.5.0

### August 2023
- implement  'tol' argument for nsev functions
- implement optional 'bb' (bounding box) argument for nsev functions
- implement optional 'gs' (grid_spacing) argument for kdvv
- added tests and additions from FNFT branch nsep-improvements, merged into development
- general:
  - updated some docstrings
  - add more tests
- **manakovv**: 
  - added interface to the FNFTs manakovv functions

- **kdvv**:
  - following the update if FNFTs kdvv,
    - kdvv now has more options, bound states and discrete spectrum can be calculated
    - initial guesses for bound states can be provided, for newton bound state localization

- **nsev**:
  - implemented that initial guesses can be provided for newton bound state localization
  
## 0.4.1

### 17.01.2022

- introduced an optional argument `display_c_msg` for the main functions `kdvv`, `kdvv_wrapper`, `nsep`, `nsep_wrapper`,
  `nsev`, `nsev_wrapper`, `nsev_inverse`, `nsev_inverse_wrapper` and `nsev_inverse_xi_wrapper`. 
  Default it is set to `true`. If set to `false`, messages from the c-library to stdout are suppressed.

### 28.06.2021
- changed the way how ctypes.CDLL is called.
- Some users (WIN) had problems to load the libfnft.dll file although the proper path was given in _get_lib_path.
  The (somewhat bad) solution introduced is to pass an `winmode=` argument to the CDLL call. It showed that passing '0'
  fixes the bug, so I introduced the function `get_winmode_param()` in `auxiliary.py` which as a standard returns 0.
  If you encounter problems, try setting the return value of this function e.g. to 'None'.


### 08.07.2020 checked for compatibility with FNFT 0.40
- kdvv: updated available discretizations in documentation, phase_shift option available
- nsep: updated available discretizations in documentation
- nsev: updated available discretizations in documentation, richardson extrapolation flag available
- nsev_inverse: updated available discretization in documentation


## 0.3.0

### 07.03.2020
- checked compatibility with FNFT 0.3.0
- merge development into master

## 0.2.2

### 05.03.2020
- adapted small changes in NSEP options/API from FNFT

### 31.10.2019

- introduced setup.py for installation with pip


### 23.12.2018

- checked compliance with FNFT 0.2.2: no changes in code

## 0.2.1

## 19.05.2019
- in dev: numerical value of nsep test output changed slightly - adapted test

### 21.11.2018
- small changes in documentation
- small changes in examples shown in README
  
### 20.11.2018
- moved example functions into separate folder
- added testfunctions to test basic interaction with FNFT
- refactoring according to PEP8
- option structs now have unified __repr__ and __print__ functions

### 08.11.2018
- fix check_return_code output warning
- re-structuring of nsev_inverse_wrapper decision tree (bound states given? contspec given?).
    Should show no effect on existing code and prevents Python crash when none of both is given (FNFT will return 7 and q all zeros in that case).

### 02.11.2018
- nsev, nsev_wrapper:
    - options switch: dst=3 now allows skipping the calculation of the discrete spectrum
    - options switch: cst=3 now allows skipping the calculation of the continuous spectrum
    - order of cst/dst options handling
- nsev_inverse
    - NONE can be passed as continuous spectrum for pure soliton states
- nsev_inverse_wrapper
    - M=0 can be passed to omit continuous spectrum
    
### 15.10.2018
 - small changes in README

### 01.10.2018
- updated nsev_inverse_example to mimic the C example

### 28.09.2018 
- added changelog
- no changes, checked compatibility with FNFT 0.2.1

examples/kdvv_examples.py
"""
This file is part of FNFTpy.
FNFTpy provides wrapper functions to interact with FNFT,
a library for the numerical computation of nonlinear Fourier transforms.

For FNFTpy to work, a copy of FNFT has to be installed.
For general information, source files and installation of FNFT,
visit FNFT's github page: https://github.com/FastNFT

For information about setup and usage of FNFTpy see README.md or documentation.

FNFTpy is free software; you can redistribute it and/or
modify it under the terms of the version 2 of the GNU General
Public License as published by the Free Software Foundation.

FNFTpy is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program. If not, see <http://www.gnu.org/licenses/>.

Contributors:

Christoph Mahnke, 2018-2023

"""

from FNFTpy import kdvv, cmplxrpr
import numpy as np


def kdvv_example(dis=None, bsl=None, bsg=None, niter=None, dst=None, cst=None, nf=None, gs=None,
                 ref=None, K=None, amplitude_scale=1.0, verbose=True, display_c_msg=True):
    """Mimics the C example for calling fnft_kdvv, can be modified with options"""
    if verbose:
        print("\n\nkdvv example")
    # set values
    D = 256
    tvec = np.linspace(-1, 1, D)
    q = np.zeros(D, dtype=np.complex128)
    q[:] = amplitude_scale * 2.0 + 0.0j
    if gs is None:
        gs = np.sqrt( np.max( np.abs(q) ) ) / 1000.  # grid spacing
    Xi1 = -2
    Xi2 = 2
    M = 8
    if K is None:
        K = D
    Xivec = np.linspace(Xi1, Xi2, M)
    # call function
    res = kdvv(q, tvec, K, M, Xi1=Xi1, Xi2=Xi2,
               dis=dis, bsl=bsl, niter=niter, dst=dst, cst=cst, nf=nf,
               ref=ref, bsg=bsg, gs = gs, display_c_msg=display_c_msg)
    # print results
    if verbose:
        print("\n----- options used ----")
        print(res['options'])
        print("\n------ results --------")
        print("FNFT return value: %d (should be 0)" % res['return_value'])
        print("continuous spectrum: ")
        for i in range(len(res['cont_ref'])):
            print("%d : Xi=%.4f   %s" % (i, Xivec[i],
                                         cmplxrpr(res['cont_ref'][i], formatter='f')))
        print("discrete spectrum:")
        print("bound state at %s  with norming constant %s" % (
            cmplxrpr(res['bound_states'][0], formatter='f'),
            cmplxrpr(res['disc_norm'][0], formatter='f')))
    return res



def kdvv_example_mex4(log2D, dis=0, diskey="", verbose=True):
    """calculation of bound states, similar to mex example 4

    Exact values of the bound states

    bound_states_exact = [1i,3i];
    normconsts_exact = [-9e6, 729e18];
    """
    T1 = 0;
    T2 = 16
    Xi1 = 0.5;
    Xi2 = 23
    tvec = np.linspace(T1, T2, 2 ** log2D)
    xivec = np.linspace(Xi1, Xi2, 2 ** log2D)
    A = 15.0
    d = 0.5
    exp_t0 = 3000.
    q = A * (1. / np.cosh((tvec - np.log(exp_t0)) / d)) ** 2
    dxi = np.sqrt( np.max( np.abs(q) ) ) / 1000.
    res = kdvv(q, tvec, M=2 ** log2D, Xi1=Xi1, Xi2=Xi2, cst=4, dis=dis, bsl=1, niter=20, gs=dxi)
    if verbose:
        print("dis=%d (%s) ... bound states: " % (dis, diskey), res['bound_states'], "  norm const", res['disc_norm'])
    return res
